tests/test_mesh.py

```diff
@@ -290,26 +291,98 @@ def test_meshio_cycle(m):
     assert_array_equal(M.t, m.t)
 
 
+_test_lambda = {
+    'left': lambda x: x[0] < 0.5,
+    'right': lambda x: x[0] > 0.5,
+}
+
+
 @pytest.mark.parametrize(
     "m",
     [
         MeshTri(),
         MeshQuad(),
+        # MeshHex(),  # TODO facet order changes?
         MeshTet(),
+    ]
+)
+def test_meshio_cycle_boundaries(m):
+
+    m = m.with_boundaries(_test_lambda)
+    M = from_meshio(to_meshio(m, write_boundaries=True))
+    assert_array_equal(M.p, m.p)
+    assert_array_equal(M.t, m.t)
+    for key in m.boundaries:
+        assert_array_equal(M.boundaries[key], m.boundaries[key])
+
+
+@pytest.mark.parametrize(
+    "m",
+    [
+        MeshTri(),
+        MeshQuad(),
         MeshHex(),
+        MeshTet(),
     ]
 )
-def test_saveload_cycle(m):
+def test_meshio_cycle_subdomains(m):
+
+    m = m.refined().with_subdomains(_test_lambda)
+    M = from_meshio(to_meshio(m))
+    assert_array_equal(M.p, m.p)
+    assert_array_equal(M.t, m.t)
+    for key in m.subdomains:
+        assert_array_equal(M.subdomains[key], m.subdomains[key])
+
+
+@pytest.mark.parametrize(
+    "m",
+    [
+        MeshTri(),
+        MeshQuad(),
+        MeshTet(),
+        MeshHex(),
+    ]
+)
+def test_saveload_cycle_vtk(m):
 
     from tempfile import NamedTemporaryFile
     m = m.refined(2)
-    f = NamedTemporaryFile(delete=False)
-    m.save(f.name + ".vtk")
-    with pytest.warns(UserWarning):
-       m2 = Mesh.load(f.name + ".vtk")
+    with NamedTemporaryFile() as f:
+        m.save(f.name + ".vtk")
+        with pytest.warns(UserWarning):
+            m2 = Mesh.load(f.name + ".vtk")
+
+    assert_array_equal(m.p, m2.p)
+    assert_array_equal(m.t, m2.t)
+
+
+@pytest.mark.parametrize(
+    "m",
+    [
+        MeshTri(),
+        MeshQuad(),
+        # MeshHex(),  # TODO facet order changes?
+        MeshTet(),
+    ]
+)
+def test_saveload_cycle_tags(m):
+
+    m = (m
+         .refined()
+         .with_subdomains(_test_lambda)
+         .with_boundaries({'test': lambda x: x[0] == 0}))
+    from tempfile import NamedTemporaryFile
+    with NamedTemporaryFile() as f:
+        m.save(f.name + ".inp", write_boundaries=True)
+        m2 = Mesh.load(f.name + ".inp")
```

Comment:
This is funny that Abaqus inp format works properly for write/read cycle but almost nothing else does without sorcery.